tools/src/wyvern/tools/parsing/coreparser/arch/DeclCheckVisitor.java

```diff
@@ -0,0 +1,230 @@
+package wyvern.tools.parsing.coreparser.arch;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+
+import wyvern.target.corewyvernIL.support.InterpreterState;
+import wyvern.tools.errors.ErrorMessage;
+import wyvern.tools.errors.ToolError;
+
+public class DeclCheckVisitor extends ArchParserDefaultVisitor {
+    private InterpreterState state;
+    private HashSet<String> connectorTypes = new HashSet<>();
+    // "componentTypes" maps component type name to the component type declaration AST node
+    private HashMap<String, ASTComponentTypeDecl> componentTypes = new HashMap<>();
+    // components maps component instance name to the component declaration AST node
+    private HashMap<String, ASTComponentDecl> components = new HashMap<>();
+    // "connectors" maps connector instance name to the connector type
+    private HashMap<String, String> connectors = new HashMap<>();
+    // entrypoints maps entrypoint name to the corresponding component instance name
+    private HashMap<String, String> entrypoints = new HashMap<>();
+    // attachments maps connector instance name to the component ports in the connection
+    private HashMap<String, HashSet<String>> attachments = new HashMap<>();
+    // portdecls maps port name to the port declaration AST node
+    private HashMap<String, ASTPortDecl> portdecls = new HashMap<>();
+
+    public List<ASTComponentDecl> generateDependencyGraph() {
+        // Assemble the dependency graph!
+        HashMap<String, DependencyGraphNode> nodes = new HashMap<>();
+        ArrayList<DependencyGraphNode> noReqNodes = new ArrayList<>();
+        List<ASTComponentDecl> ordered = new LinkedList<>();
+        for (String connector : attachments.keySet()) {
+            HashSet<String> fullports = attachments.get(connector);
+            for (String fullport : fullports) {
+                String component = fullport.substring(0, fullport.indexOf("."));
+                ASTComponentDecl compObj = components.get(component);
+                ASTComponentTypeDecl compType = componentTypes.get(compObj.getType());
+                if (!nodes.containsKey(component)) {
+                    DependencyGraphNode compNode = new DependencyGraphNode(compObj);
+                    nodes.put(component, compNode);
+                }
+                DependencyGraphNode compNode = nodes.get(component);
+                if (compType.getReqs().size() == 0) {
+                    noReqNodes.add(compNode);
+                }
```

Comment:
It seems to be that this should be done outside the loop after line 70. At the time the dependency graph node is created, its requires list is going to be empty. So, it would be added to noReqNodes even though it should not be added.

Reply:
In light of the fact that I would have to fix this function later, I guess it's okay if you leave lines 45-47 here.